### PR TITLE
Make TvdbClientManager thread safe

### PR DESCRIPTION
**Pull request by Spacetech**
Along with #4222, another issue I ran into was when multiple `TvdbClientManager` calls ran concurrently.

Here's some errors that will happen if you try that:
```
[2020-09-27 15:27:00.698 -07:00] [ERR] [16] MediaBrowser.Providers.TV.SeriesMetadataService: Error in "TheTVDB"
System.ArgumentException: An item with the same key has already been added. Key: System.Net.Http.Headers.HeaderDescriptor
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Net.Http.Headers.HttpHeaders.GetOrCreateHeaderInfo(HeaderDescriptor descriptor, Boolean parseRawValues)
   at System.Net.Http.Headers.HttpHeaders.SetParsedValue(HeaderDescriptor descriptor, Object value)
   at System.Net.Http.Headers.HttpRequestHeaders.set_Authorization(AuthenticationHeaderValue value)
   at TvDbSharper.Clients.AuthenticationClient.UpdateAuthenticationHeader(String token)
   at TvDbSharper.Clients.AuthenticationClient.AuthenticateAsync(AuthenticationData authenticationData, CancellationToken cancellationToken)
   at MediaBrowser.Providers.Plugins.TheTvdb.TvdbClientManager.GetTvDbClient(String language) in C:\Programming\jellyfin\MediaBrowser.Providers\Plugins\TheTvdb\TvdbClientManager.cs:line 49
```
```
[2020-09-27 14:36:38.036 -07:00] [ERR] [44] MediaBrowser.Providers.TV.SeriesMetadataService: Error in "TheTVDB"
System.ArgumentException: An item with the same key has already been added. Key: System.Net.Http.Headers.HeaderDescriptor
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at System.Net.Http.Headers.HttpHeaders.Add(HeaderDescriptor descriptor, String value)
   at System.Net.Http.Headers.HttpHeaderValueCollection`1.ParseAdd(String input)
   at TvDbSharper.TvDbClient.set_AcceptedLanguage(String value)
   at MediaBrowser.Providers.Plugins.TheTvdb.TvdbClientManager.TryGetValue[T](String key, String language, Func`1 resultFactory) in C:\Programming\jellyfin\MediaBrowser.Providers\Plugins\TheTvdb\TvdbClientManager.cs:line 279
   at MediaBrowser.Providers.Plugins.TheTvdb.TvdbSeriesProvider.FindSeriesInternal(String name, String language, CancellationToken cancellationToken) in C:\Programming\jellyfin\MediaBrowser.Providers\Plugins\TheTvdb\TvdbSeriesProvider.cs:line 229
   at MediaBrowser.Providers.Plugins.TheTvdb.TvdbSeriesProvider.FindSeries(String name, Nullable`1 year, String language, CancellationToken cancellationToken) in C:\Programming\jellyfin\MediaBrowser.Providers\Plugins\TheTvdb\TvdbSeriesProvider.cs:line 197
   at MediaBrowser.Providers.Plugins.TheTvdb.TvdbSeriesProvider.Identify(SeriesInfo info) in C:\Programming\jellyfin\MediaBrowser.Providers\Plugins\TheTvdb\TvdbSeriesProvider.cs:line 398
   at MediaBrowser.Providers.Plugins.TheTvdb.TvdbSeriesProvider.GetMetadata(SeriesInfo itemId, CancellationToken cancellationToken) in C:\Programming\jellyfin\MediaBrowser.Providers\Plugins\TheTvdb\TvdbSeriesProvider.cs:line 80
   at MediaBrowser.Providers.Manager.MetadataService`2.ExecuteRemoteProviders(MetadataResult`1 temp, String logName, TIdType id, IEnumerable`1 providers, CancellationToken cancellationToken) in C:\Programming\jellyfin\MediaBrowser.Providers\Manager\MetadataService.cs:line 853
```

It's complaining about setting `AcceptedLanguage` and calling `AuthenticateAsync`.

A fix is to have individual clients per language, so `AcceptedLanguage` is set once during the initial creation. Along with that, add a locking mechanism to prevent concurrent authentication requests.

**Changes**
- Keep a `ConcurrentDictionary` of `TvDbClient`s per language.
- Add a locking mechanism to prevent concurrent authentication requests. Technically we could introduce a lock per language so clients could authenticate concurrently but I doubt that would result in a notice performance increase.
- Change `GetTvDbClient` to return a specific langauges `TvDbClient` and make it async so the token methods can be await'd
- Update `TryGetValue` to use `GetOrCreateAsync`. Something to note is that `MemoryCache.GetOrCreateAsync` is thread safe but has odd behavior ([ref 1](https://blog.novanet.no/asp-net-core-memory-cache-is-get-or-create-thread-safe/), [ref 2](https://www.hanselman.com/blog/EyesWideOpenCorrectCachingIsAlwaysHard.aspx)). So it might run the factory a few times for the same key at the start. This is kind of annoying but might not be a huge issue for this use case.
- Use `UtcNow` for the token `DateTime` since it's less scary than local times.

**Issues**
None

### CODE REVIEW

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 1]
```diff
@@ -269,17 +297,18 @@ public async IAsyncEnumerable<KeyType> GetImageKeyTypesForSeasonAsync(int tvdbId
...
         }
 
-        private async Task<T> TryGetValue<T>(string key, string language, Func<Task<T>> resultFactory)
+        private Task<T> TryGetValue<T>(string key, string language, Func<TvDbClient, Task<T>> resultFactory)
         {
-            if (_cache.TryGetValue(key, out T cachedValue))
-            {
-                return cachedValue;
-            }
+            return _cache.GetOrCreateAsync(key, async (entry) =>
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:302]:
```suggestion
            return _cache.GetOrCreateAsync(key, async entry =>
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 2]
```diff
@@ -254,7 +282,7 @@ public async IAsyncEnumerable<KeyType> GetImageKeyTypesForSeriesAsync(int tvdbId
         public async IAsyncEnumerable<KeyType> GetImageKeyTypesForSeasonAsync(int tvdbId, string language, [EnumeratorCancellation] CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey(nameof(TvDbClient.Series.GetImagesSummaryAsync), tvdbId);
-            var imagesSummary = await TryGetValue(cacheKey, language, () => TvDbClient.Series.GetImagesSummaryAsync(tvdbId, cancellationToken)).ConfigureAwait(false);
+            var imagesSummary = await TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Series.GetImagesSummaryAsync(tvdbId, cancellationToken)).ConfigureAwait(false);
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:285]:
```suggestion
            var imagesSummary = await TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Series.GetImagesSummaryAsync(tvdbId, cancellationToken)).ConfigureAwait(false);
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 3]
```diff
@@ -233,7 +261,7 @@ await GetEpisodesPageAsync(Convert.ToInt32(seriesTvdbId), episodeQuery, language
         public async IAsyncEnumerable<KeyType> GetImageKeyTypesForSeriesAsync(int tvdbId, string language, [EnumeratorCancellation] CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey(nameof(TvDbClient.Series.GetImagesSummaryAsync), tvdbId);
-            var imagesSummary = await TryGetValue(cacheKey, language, () => TvDbClient.Series.GetImagesSummaryAsync(tvdbId, cancellationToken)).ConfigureAwait(false);
+            var imagesSummary = await TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Series.GetImagesSummaryAsync(tvdbId, cancellationToken)).ConfigureAwait(false);
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:264]:
```suggestion
            var imagesSummary = await TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Series.GetImagesSummaryAsync(tvdbId, cancellationToken)).ConfigureAwait(false);
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 4]
```diff
@@ -166,9 +196,7 @@ public Task<TvDbResponse<Language[]>> GetLanguagesAsync(CancellationToken cancel
             CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey(language, tvdbId, episodeQuery);
-
-            return TryGetValue(cacheKey, language,
-                () => TvDbClient.Series.GetEpisodesAsync(tvdbId, page, episodeQuery, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Series.GetEpisodesAsync(tvdbId, page, episodeQuery, cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:199]:
```suggestion
            return TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Series.GetEpisodesAsync(tvdbId, page, episodeQuery, cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 5]
```diff
@@ -154,8 +185,7 @@ public Task<TvDbResponse<Language[]>> GetLanguagesAsync(CancellationToken cancel
             CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey("seriesepisodesummary", tvdbId, language);
-            return TryGetValue(cacheKey, language,
-                () => TvDbClient.Series.GetEpisodesSummaryAsync(tvdbId, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Series.GetEpisodesSummaryAsync(tvdbId, cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:188]:
```suggestion
            return TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Series.GetEpisodesSummaryAsync(tvdbId, cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 6]
```diff
@@ -140,12 +171,12 @@ private TvDbClient TvDbClient
...
         {
             var cacheKey = GenerateKey("images", tvdbId, language, imageQuery);
-            return TryGetValue(cacheKey, language, () => TvDbClient.Series.GetImagesAsync(tvdbId, imageQuery, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Series.GetImagesAsync(tvdbId, imageQuery, cancellationToken));
         }
 
         public Task<TvDbResponse<Language[]>> GetLanguagesAsync(CancellationToken cancellationToken)
         {
-            return TryGetValue("languages", null, () => TvDbClient.Languages.GetAllAsync(cancellationToken));
+            return TryGetValue("languages", null, (tvDbClient) => tvDbClient.Languages.GetAllAsync(cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:179]:
```suggestion
            return TryGetValue("languages", null, tvDbClient => tvDbClient.Languages.GetAllAsync(cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 7]
```diff
@@ -140,12 +171,12 @@ private TvDbClient TvDbClient
             CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey("images", tvdbId, language, imageQuery);
-            return TryGetValue(cacheKey, language, () => TvDbClient.Series.GetImagesAsync(tvdbId, imageQuery, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Series.GetImagesAsync(tvdbId, imageQuery, cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:174]:
```suggestion
            return TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Series.GetImagesAsync(tvdbId, imageQuery, cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 8]
```diff
@@ -130,7 +161,7 @@ private TvDbClient TvDbClient
             CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey("actors", tvdbId, language);
-            return TryGetValue(cacheKey, language, () => TvDbClient.Series.GetActorsAsync(tvdbId, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Series.GetActorsAsync(tvdbId, cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:164]:
```suggestion
            return TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Series.GetActorsAsync(tvdbId, cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 9]
```diff
@@ -121,7 +152,7 @@ private TvDbClient TvDbClient
             CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey("series", zap2ItId, language);
-            return TryGetValue(cacheKey, language, () => TvDbClient.Search.SearchSeriesByZap2ItIdAsync(zap2ItId, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Search.SearchSeriesByZap2ItIdAsync(zap2ItId, cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:155]:
```suggestion
            return TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Search.SearchSeriesByZap2ItIdAsync(zap2ItId, cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 10]
```diff
@@ -112,7 +143,7 @@ private TvDbClient TvDbClient
             CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey("series", imdbId, language);
-            return TryGetValue(cacheKey, language, () => TvDbClient.Search.SearchSeriesByImdbIdAsync(imdbId, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Search.SearchSeriesByImdbIdAsync(imdbId, cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:146]:
```suggestion
            return TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Search.SearchSeriesByImdbIdAsync(imdbId, cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 11]
```diff
@@ -21,67 +22,97 @@ public class TvdbClientManager
...
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Series.GetAsync(tvdbId, cancellationToken));
         }
 
-        public Task<TvDbResponse<EpisodeRecord>> GetEpisodesAsync(int episodeTvdbId, string language,
-            CancellationToken cancellationToken)
+        public Task<TvDbResponse<EpisodeRecord>> GetEpisodesAsync(int episodeTvdbId, string language, CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey("episode", episodeTvdbId, language);
-            return TryGetValue(cacheKey, language, () => TvDbClient.Episodes.GetAsync(episodeTvdbId, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Episodes.GetAsync(episodeTvdbId, cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:112]:
```suggestion
            return TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Episodes.GetAsync(episodeTvdbId, cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 12]
```diff
@@ -21,67 +22,97 @@ public class TvdbClientManager
...
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Search.SearchSeriesByNameAsync(name, cancellationToken));
         }
 
-        public Task<TvDbResponse<Series>> GetSeriesByIdAsync(int tvdbId, string language,
-            CancellationToken cancellationToken)
+        public Task<TvDbResponse<Series>> GetSeriesByIdAsync(int tvdbId, string language, CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey("series", tvdbId, language);
-            return TryGetValue(cacheKey, language, () => TvDbClient.Series.GetAsync(tvdbId, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Series.GetAsync(tvdbId, cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:106]:
```suggestion
            return TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Series.GetAsync(tvdbId, cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 13]
```diff
@@ -21,67 +22,97 @@ public class TvdbClientManager
...
+            return tvDbClient;
         }
 
-        public Task<TvDbResponse<SeriesSearchResult[]>> GetSeriesByNameAsync(string name, string language,
-            CancellationToken cancellationToken)
+        public Task<TvDbResponse<SeriesSearchResult[]>> GetSeriesByNameAsync(string name, string language, CancellationToken cancellationToken)
         {
             var cacheKey = GenerateKey("series", name, language);
-            return TryGetValue(cacheKey, language, () => TvDbClient.Search.SearchSeriesByNameAsync(name, cancellationToken));
+            return TryGetValue(cacheKey, language, (tvDbClient) => tvDbClient.Search.SearchSeriesByNameAsync(name, cancellationToken));
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:100]:
```suggestion
            return TryGetValue(cacheKey, language, tvDbClient => tvDbClient.Search.SearchSeriesByNameAsync(name, cancellationToken));
```

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 14]
```diff
@@ -21,67 +22,97 @@ public class TvdbClientManager
...
+                try
                 {
-                    try
+                    await _tokenUpdateLock.WaitAsync().ConfigureAwait(false);
+
+                    if (string.IsNullOrEmpty(tvDbClient.Authentication.Token))
                     {
-                        _tvDbClient.Authentication.RefreshTokenAsync().GetAwaiter().GetResult();
+                        await tvDbClient.Authentication.AuthenticateAsync(TvdbUtils.TvdbApiKey).ConfigureAwait(false);
+                        _tokenUpdatedAt = DateTime.UtcNow;
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:58]: Might have issues with auth tokens if the clients aren't created at the same time

Comment (Spacetech) [reply MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs]: Good call out. I refactored it a bit so there's a separate token time/lock per client

**Review by cvium — APPROVED** [review]

File: MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs [hunk 15]
```diff
@@ -21,67 +22,93 @@ public class TvdbClientManager
...
+            // First time authenticating if the token was never updated or if it's empty in the client
+            if (tvDbClientInfo.TokenUpdatedAt == DateTime.MinValue || string.IsNullOrEmpty(tvDbClient.Authentication.Token))
             {
-                if (string.IsNullOrEmpty(_tvDbClient.Authentication.Token))
+                try
                 {
-                    _tvDbClient.Authentication.AuthenticateAsync(TvdbUtils.TvdbApiKey).GetAwaiter().GetResult();
-                    _tokenCreatedAt = DateTime.Now;
-                }
+                    await tvDbClientInfo.TokenUpdateLock.WaitAsync().ConfigureAwait(false);
```
Comment (Bond-009) [inline MediaBrowser.Providers/Plugins/TheTvdb/TvdbClientManager.cs:49]: This should be on the outside of the try block

**Review by Bond-009 — CHANGES_REQUESTED** [review]

**Review by crobibero — APPROVED** [review]

**Comment by crobibero** [comment]
The TVDB plugin has been removed from server source, if we still want this please create a PR in that repo (https://github.com/jellyfin/jellyfin-plugin-tvdb)

**Pull request closed without merging**